integration_test/cases/type.exs

```diff
@@ -141,6 +142,11 @@ defmodule Ecto.Integration.TypeTest do
     # Comparison expression
     assert [12] = TestRepo.all(from p in Post, select: type(coalesce(p.visits, 0), :integer))
     assert [1.0] = TestRepo.all(from p in Post, select: type(coalesce(p.intensity, 1.0), :float))
+
+    # parent_as/1
+    child = from c in Comment, where: type(parent_as(:posts).id, :string) == c.text, select: c.post_id
+    query = from p in Post, as: :posts, where: p.id in subquery(child), select: p.id
+    assert [post_id] == TestRepo.all(query)
```

Comment:
I don't think we need the integration test in this case as the restriction is fully on Ecto.Query side, but since you wrote it already, we can go ahead and commit it. :)